reference/matrix/csr_kernels.cpp

```diff
@@ -396,6 +397,80 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_CSR_CALCULATE_MAX_NNZ_PER_ROW_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void convert_to_hybrid(std::shared_ptr<const ReferenceExecutor> exec,
+                       matrix::Hybrid<ValueType, IndexType> *result,
+                       const matrix::Csr<ValueType, IndexType> *source)
+{
+    auto num_rows = result->get_size()[0];
+    auto num_cols = result->get_size()[1];
+    auto strategy = result->get_strategy();
+    auto ell_lim = strategy->get_ell_num_stored_elements_per_row();
+    auto coo_lim = strategy->get_coo_nnz();
+    auto coo_val = result->get_coo_values();
+    auto coo_col = result->get_coo_col_idxs();
+    auto coo_row = result->get_coo_row_idxs();
+
+    // Initial Hybrid Matrix
+    for (size_type i = 0; i < result->get_ell_num_stored_elements_per_row();
+         i++) {
+        for (size_type j = 0; j < result->get_ell_stride(); j++) {
+            result->ell_val_at(j, i) = zero<ValueType>();
+            result->ell_col_at(j, i) = 0;
+        }
+    }
+    for (size_type i = 0; i < result->get_coo_num_stored_elements(); i++) {
+        coo_val[i] = zero<ValueType>();
+        coo_col[i] = 0;
+        coo_row[i] = 0;
+    }
+
+    const auto csr_row_ptrs = source->get_const_row_ptrs();
+    const auto csr_vals = source->get_const_values();
+    size_type csr_idx = 0;
+    size_type coo_idx = 0;
+    for (IndexType row = 0; row < num_rows; row++) {
+        size_type ell_idx = 0;
+        while (csr_idx < csr_row_ptrs[row + 1]) {
+            auto val = csr_vals[csr_idx];
+            if (val != zero<ValueType>()) {
+                if (ell_idx < ell_lim) {
+                    result->ell_val_at(row, ell_idx) = val;
+                    result->ell_col_at(row, ell_idx) =
+                        source->get_const_col_idxs()[csr_idx];
+                    ell_idx++;
+                } else {
+                    coo_val[coo_idx] = val;
+                    coo_col[coo_idx] = source->get_const_col_idxs()[csr_idx];
+                    coo_row[coo_idx] = row;
+                    coo_idx++;
+                }
+            }
+            csr_idx++;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CSR_CONVERT_TO_HYBRID_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void calculate_nonzeros_per_row(std::shared_ptr<const ReferenceExecutor> exec,
+                                const matrix::Csr<ValueType, IndexType> *source,
+                                Array<size_type> *result)
+{
+    const auto row_ptrs = source->get_const_row_ptrs();
+    auto row_nnz_val = result->get_data();
+    for (size_type i = 0; i < result->get_num_elems(); i++) {
+        row_nnz_val[i] = row_ptrs[i + 1] - row_ptrs[i];
```

Comment:
In all other matrix formats, we test each value if it is zero, or not (to be fair, all of them are either derived from ELL or dense, where we do that). I like this optimization, but it could result in a larger nnz_per_row than actually present (when a value is actually zero).
However, in the Coo part of Hybrid in this PR, you actually do check every value of COO instead of just using `get_num_stored_elements()`: 
```
coo_nnz += (coo_val[ind] != zero<ValueType>());
```
[Taken from here](https://github.com/ginkgo-project/ginkgo/pull/302/files#diff-1ca8237f66d76e9de4003a9aedbf42ffR150)

So I think it should be dealt with uniformly.

What do you think, @hartwiganzt , should we count the actual NNZ here, or can we rely on the `row_ptrs` for that? It is kind of a performance question, should the conversion be very optimized, or can it take a bit longer, which might result in lower memory consumption (matrix performance will not be impacted since during the conversion, every value is checked before it is written into the other format)?
I am a fan of this optimization (and would lean towards keeping this in here), however, if the user (or some algorithm) deletes parts of the values (sets them to zero), we might allocate too much memory. Admittedly, in the conversion from COO to CSR (and vice versa), we also ignore that since we can reuse the value and column pointer (so we don't have to check the actual values).

Reply:
Difficult. Sometimes you want to store some explicit zeros also in the sparse format to preserve coalesced memory access and cache reuse. An example are stencil matrices. Assume you store in ELL format, then you want to explicitly store also zero values to have the access to the vector values fully coalesced. This is why I lean towards keeping it as is.

Reply:
Then I guess we leave it as it is right now.
The COO part of Hybrid should be fine then since we should not have any cache advantages there (because they contain only the lines with too many elements).

I will mark this as resolved then.